src/java.base/share/classes/java/lang/classfile/AnnotationValue.java

```diff
@@ -432,9 +432,52 @@ default ClassDesc classSymbol() {
         Utf8Entry constantName();
     }
 
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfByte}. */
+    int TAG_BYTE = 'B';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfChar}. */
+    int TAG_CHAR = 'C';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfDouble}. */
+    int TAG_DOUBLE = 'D';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfFloat}. */
+    int TAG_FLOAT = 'F';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfInt}. */
+    int TAG_INT = 'I';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfLong}. */
+    int TAG_LONG = 'J';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfShort}. */
+    int TAG_SHORT = 'S';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfBoolean}. */
+    int TAG_BOOLEAN = 'Z';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfString}. */
+    int TAG_STRING = 's';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfEnum}. */
+    int TAG_ENUM = 'e';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfClass}. */
+    int TAG_CLASS = 'c';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfAnnotation}. */
+    int TAG_ANNOTATION = '@';
+
+    /** The {@link #tag() tag} indicating the value of an element-value pair is {@link OfArray}. */
+    int TAG_ARRAY = '[';
+
     /**
      * {@return the tag character for this value as per JVMS {@jvms 4.7.16.1}}
      * The tag characters have a one-to-one mapping to the types of annotation element values.
+     *
+     * @apiNote
+     * {@code TAG_}-prefixed constants in this class, such as {@link #TAG_INT},
+     * describe the possible return values of this method.
      */
     char tag();
```

Comment:
> First, some bit of history. These API methods are added before those constants, and this patch did not change the type of these methods or constants.

Sure, makes sense to separate this patch from a potential type change of the methods/constants.

> I believe the methods have restricted return types to be informative: `tag()` is a U1, but it can be interpreted as a char (see [JVMS 4.7.16.1](https://docs.oracle.com/javase/specs/jvms/se23/html/jvms-4.html#jvms-4.7.16.1)) and thus is returned as one. CP tag is a U1 too (see [JVMS 4.4](https://docs.oracle.com/javase/specs/jvms/se23/html/jvms-4.html#jvms-4.7.16.1)) so maybe `byte` is just used out of convenience.

I think `byte` and `char` as the types are a good choice given the way these tag items are described in the JVMS.

> However, we usually use `int` to represent U1, also called unsigned bytes. See [`DataInput::readUnsignedByte`](https://github.com/openjdk/jdk/blob/c43202baf6eb7e49ec458037971a9efa392d053e/src/java.base/share/classes/java/io/DataInput.java#L323) and [`ClassReader::readU1`](https://github.com/openjdk/jdk/blob/c43202baf6eb7e49ec458037971a9efa392d053e/src/java.base/share/classes/java/lang/classfile/ClassReader.java#L133). Guess that's why we used `int` for the constants. (Also, note that byte and char are all [erased to int](https://download.java.net/java/early_access/jdk24/docs/api/java.base/java/lang/classfile/TypeKind.html#computational-type) in bytecode, so this has little practical effect and that might be why I didn't bother to change these)

One scenario where the unequal types of the methods and constants _does_ have a practical effect are JVM languages that don't have [binary numeric promotion](https://docs.oracle.com/javase/specs/jls/se23/html/jls-5.html#jls-5.6) for their [numerical equality operators](https://docs.oracle.com/javase/specs/jls/se23/html/jls-15.html#jls-15.21.1).

This is the case for Kotlin's [value equality expressions](https://kotlinlang.org/spec/expressions.html#value-equality-expressions). This Kotlin code [would not compile](https://pl.kotl.in/A9D1WDdS8):
```kotlin
fun isArray(a: AnnotationValue): Boolean {
    return a.tag() == AnnotationValue.TAG_ARRAY
}
```
Instead you would have to write [this code](https://pl.kotl.in/qSrwSIL39):
```kotlin
fun isArray(a: AnnotationValue): Boolean {
    return a.tag() == AnnotationValue.TAG_ARRAY.toChar()
}
```

Reply:
Your example is an exact antipattern from our data-oriented model: we would want users to check the object type with `instanceof` (should be `is` in kotlin) instead of checking these constants.

Yet I think we can consider promoting Constant Pool tag from byte or char, short, or int to represent a u1 in case it goes over 127.

Reply:
> Your example is an exact antipattern from our data-oriented model: we would want users to check the object type with `instanceof` (should be `is` in kotlin) instead of checking these constants.

I'm aware, this was just the first example I could come up with.

> Yet I think we can consider promoting Constant Pool tag from byte or char, short, or int to represent a u1 in case it goes over 127.

Maybe just using `int` everywhere for consistency might be a good option too.

Reply:
> Yet I think we can consider promoting Constant Pool tag from byte or char, short, or int to represent a u1 in case it goes over 127.

Is there any chance a change like this could make it into JDK 24? I'd imagine it would be too late after that because it's a binary incompatible change.

Reply:
Don't think so. This imo is fixable  in the future if we do have many new cp/annotation tags (see example of Thread::threadId), though I don't think that is likely

Reply:
Wait, I referenced the wrong thing. I was speaking about this:
> Maybe just using `int` everywhere for consistency might be a good option too.

Reply:
Can you try raise this problem in this mailing list https://mail.openjdk.org/mailman/listinfo/classfile-api-dev?
I don't think this is critical enough from a Java developer's perspective and we are close to RDP 1, but people on the list might agree with your proposal and we might get an emergency patch (with CSR reviews) to fix this up.

Reply:
> Can you try raise this problem in this mailing list https://mail.openjdk.org/mailman/listinfo/classfile-api-dev?

Thanks for [mentioning it on the list](https://mail.openjdk.org/pipermail/classfile-api-dev/2024-December/000613.html) and implementing it in #22489! I did not find the time yet to post myself. :)